### PR TITLE
Add a `config.openshift.io/Infrastructure` instance to the cluster

### PR DESCRIPTION
**Pull request by ironcladlou**
Add a `config.openshift.io/Infrastructure` instance to the cluster.

* Update the `openshift/api` package to a version containing the new config type.
* Generate the `config.openshift.io/Infrastructure` custom resource definition manifest.
* Generate the CRD instance manifest populated with cloud provider name from the install config.

This enables consumers of cloud provider configuration to migrate away from install config.

/cc @openshift/sig-network-edge @rajatchopra @deads2k

### CODE REVIEW

File: pkg/asset/manifests/infrastructure.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	return nil
+}
+
+// Files returns the files generated by the asset.
+func (i *Infrastructure) Files() []*asset.File {
+	return i.FileList
+}
+
+// Load loads the already-rendered files back from disk.
+func (i *Infrastructure) Load(f asset.FileFetcher) (bool, error) {
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply pkg/asset/manifests/infrastructure.go]: I'll refactor DNS once this PR merges

File: pkg/asset/manifests/infrastructure.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+var (
+	infraCrdFilename = "cluster-infrastructure-01-crd.yaml"
+	infraCfgFilename = filepath.Join(manifestDir, "cluster-infrastructure-02-config.yml")
+)
+
+// Infrastructure generates the cluster-infrastructure-*.yml files.
+type Infrastructure struct {
+	config   *configv1.Infrastructure
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply pkg/asset/manifests/infrastructure.go]: Fixed

File: pkg/asset/manifests/infrastructure.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	configData, err := yaml.Marshal(i.config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create %s manifests from InstallConfig", i.Name())
+	}
+
+	crdData, err := content.GetBootkubeTemplate(infraCrdFilename)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply pkg/asset/manifests/infrastructure.go]: Fixed — also, I wasn't aware of `github.com/pkg/errors`, I like it and will propose using it in my team's operators. Thanks!

File: pkg/asset/manifests/infrastructure.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+			// not namespaced
+		},
+		Status: configv1.InfrastructureStatus{
+			CloudProvider: provider,
+		},
+	}
+
+	configData, err := yaml.Marshal(i.config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create %s manifests from InstallConfig", i.Name())
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply pkg/asset/manifests/infrastructure.go]: Fixed

**Comment by abhinavdahiya** [comment]
/hold

Cloud provider is the wrong abstraction in the config. All cloudproviders are going to be `external` pretty soon. So I rather see Platform been set here...

**Comment by ironcladlou** [comment]
@abhinavdahiya 

>Cloud provider is the wrong abstraction in the config. All cloudproviders are going to be external pretty soon. So I rather see Platform been set here...

That seems like a discussion we should have had in https://github.com/openshift/api/pull/145. Are you going to open a new API PR with some proposed change?

File: Gopkg.lock [hunk 5]
```diff
@@ -364,14 +364,14 @@
   version = "1.0.1"
 
 [[projects]]
-  digest = "1:68214731af5ff5a3bfab4d28571578e5522bc4f667ad1232745d7b4189ccb442"
+  digest = "1:400ab065329172f4f4a88ad94401795780566749b1b3cb73029d0d3971659925"
```
[the remark this answers is not in the document]
Comment (staebler) [reply Gopkg.lock]: Ah, yes, sorry for the noise.

**Comment by staebler** [comment]
@rajatchopra Is there a distinction between manifests that should be part of manifest-templates and those that should not? I am wondering whether this new Infrastructure Config asset should be output from the manifest-templates target. It is not clear to me why the Ingress Config and Network Config are not.

**Comment by rajatchopra** [comment]
> @rajatchopra Is there a distinction between manifests that should be part of manifest-templates and those that should not? I am wondering whether this new Infrastructure Config asset should be output from the manifest-templates target. It is not clear to me why the Ingress Config and Network Config are not.

@staebler I put my review comments and then saw your comment. I have the same feedback. There is no distinction as of now, I think we should put all templates/manifests as manifest-template assets first and depend upon them if needed. Unless there is a strong reason otherwise. I don't see the reason here.

**Comment by ironcladlou** [comment]
Should the `Infrastructure` instance itself also be a template like the CRD? I'm not entirely clear on what the decision making framework is for when an asset should be a Go template in vs. a manifest generated from Go code in `pkg/asset/manifests`.

**Comment by ironcladlou** [comment]
```
$ oc get infrastructures/cluster -o yaml
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
  creationTimestamp: 2019-01-09T18:53:37Z
  generation: 1
  name: cluster
  resourceVersion: "227"
  selfLink: /apis/config.openshift.io/v1/infrastructures/cluster
  uid: df61ad74-143f-11e9-a097-02040a65c016
spec: {}
status:
  platform: AWS
```

**Comment by ironcladlou** [comment]
@abhinavdahiya @rajatchopra @staebler PTAL, tests are green.

File: pkg/asset/manifests/infrastructure.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+// Generate generates the Infrastructure config and its CRD.
+func (i *Infrastructure) Generate(dependencies asset.Parents) error {
+	installConfig := &installconfig.InstallConfig{}
+	infra := &openshift.InfrastructureCRD{}
+	dependencies.Get(installConfig, infra)
+
+	var platform configv1.PlatformType
+	switch p := installConfig.Config.Platform; {
+	case p.AWS != nil:
+		platform = configv1.AWSPlatform
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/infrastructure.go]: do you think we can set the other platforms too ? like libvirt/openstack

also you can switch on `platform.Name()`

Comment (ironcladlou) [reply pkg/asset/manifests/infrastructure.go]: Done

**Comment by abhinavdahiya** [comment]
/lgtm

/hold
i'll wait for maybe @rajatchopra / @staebler be green light by the end of the day...

**Comment by wking** [comment]
[images][1] has the 504 gateway timeout.  Kicking again just in case it squeaks through:

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/943/pull-ci-openshift-installer-master-images/2043/build-log.txt

**Comment by abhinavdahiya** [comment]
> /lgtm
> 
> /hold
> i'll wait for maybe @rajatchopra / @staebler be green light by the end of the day...

/hold cancel

**Comment by ironcladlou** [comment]
@rajatchopra re-tag? I accidentally removed it

**Comment by ironcladlou** [comment]
C'mon, bot.
/retest

**Comment by Miciah** [comment]
could not copy stable imagestreamtag: Timeout: request did not complete within allowed duration

/retest

**Comment by Miciah** [comment]
level=error msg="\t* module.vpc.aws_route_table_association.worker_routing[0]: 1 error occurred:"
    level=error msg="\t* aws_route_table_association.worker_routing.0: timeout while waiting for state to become 'success' (timeout: 5m0s)"
    level=error
    level=error
    level=error msg="\t* module.vpc.aws_route_table_association.route_net[4]: 1 error occurred:"
    level=error msg="\t* aws_route_table_association.route_net.4: timeout while waiting for state to become 'success' (timeout: 5m0s)"

/retest

**Pull request merged**